[RFC] Incsub 4 (renamed to 'inccommand')

justinmk opened the pull request:
- [x] rename to 'inccommand'
- [x] back out special cases in `do_ecmd()` and `buflist_new()`
- [x] reduce shared/global state
- [x] preserve `b:changedtick` https://github.com/neovim/neovim/pull/5561#discussion_r86656183
- [x] add tests for `feedkeys(':%s/foo/bar/')`, `feedkeys(":%s/foo/bar/\<CR>")`, ... 
    - [x] check `KeyTyped` to avoid preview-mode during `feedkeys()`
- [x] should preview split appear if `:{range}` is in the viewport? (answer: yes, be predictable)
- [x] disable 'cursorline', 'spell' in preview window
- [x] set buffer name to `[Preview]` instead of `[inc_sub]`
- [x] change `[lnum]` display to `|lnum|` (like quickfix window)
- [x] highlight/multiline bug #7231 https://github.com/neovim/neovim/pull/5226#issuecomment-257153602 
- [ ] re-use _window id_ for preview buffer (`:h win_getid()`)
    - best approach here would be to _keep the window open_ and close it when command ends
- ~~should NOT change 'modified'~~ (punting this for now; too risky)
- ~~ctrl-p/ctrl-n/up/down in `:` should not trigger 'inccommand' until user edits the command~~
    - going to wait for feedback before doing this

### Renamed `'incsubstitute'` to `'inccommand'` 

So that we can expand it to other commands besides `:substitute` without adding new options (that was already implied in the implementation, re `eap->is_live`).

### Future considerations

- block RPCs/events during incremental/live preview (`eap->is_live`)?
    - the buffer is in a weird state during the preview, don't want plugins interacting with it
    - done: #9783
- ~~instead of `u_undo_and_forget()`, it might be more robust to create a special/dummy buffer and throw it away~~
    - crazy talk: to show live previews of multi-buffer commands like `:argdo`, `:cdo`, overlay a new, temporary instance of `nvim`?

KillTheMule committed:
Incsubsitution feature

Originally implemented by

* Clement0
* DesbyP
* aym7
* Adrey06
* Robinhola

in #4811. Major reworkings and bug
fixes by

* bfredl

Most tests suggested by ZyX-l, suggestions for improvements by oni-link.

KillTheMule committed:
Tests for incsubstitution feature

KillTheMule committed:
Linted

justinmk committed:
'inccommand': rework

- Eliminate/isolate static/global variables
- Remove special-case parameter from buflist_new()
- Remove special-case ECMD_RESERVED_BUFNR
- To determine when u_undo_and_forget() should be done, check
  b_changedtick instead of a heuristic.
- use mb_string2cells() instead of strlen() to measure the :sub patterns
- call ml_close() before buf_clear_file(). Avoids leaks caught by ASan.

Original patch by:
  Robin Elrharbi-Fleury (Robinhola)
  Audrey Rayé (Adrey06)
  Philémon Hullot (DesbyP)
  Aymeric Collange (aym7)
  Clément Guyomard (Clement0)

justinmk commented on src/nvim/ex_cmds.c:
```diff
@@ -6189,24 +6179,32 @@ static void incsub_display(char_u *pat, char_u *sub,
 
   emsg_off--;
+  if (save_changedtick != curbuf->b_changedtick
+      && !u_undo_and_forget(1)) {
```
`u_undo_and_forget()` does not restore `b:changedtick`, which is used by some plugins like repeat.vim. @bfredl @KillTheMule should we restore it in `u_undo_and_forget()` or would it not be appropriate there?

justinmk commented:
Only 2 tests failing, and valgrind seems OK.

justinmk committed:
lint

KillTheMule commented:
> add tests for feedkeys(':%s/foo/bar/'), feedkeys(":%s/foo/bar/\<CR>"), ...

If you could elaborate on this (I'm not sure what you mean), I could see if I can write them.

justinmk commented:
@KillTheMule Generally, to verify that "live" substitution (or any "live" handling in general) should not be invoked during anything except actual interactive `:` use.

KillTheMule commented:
Ok we have some basics: https://github.com/neovim/neovim/pull/5561/commits/561c1e39e05bac583f851de6165cd656058df7e1#diff-4dbe255ebfc94f6664c66ffeab750809R83

You mean along the lines of https://github.com/neovim/neovim/pull/5561/commits/561c1e39e05bac583f851de6165cd656058df7e1#diff-4dbe255ebfc94f6664c66ffeab750809R818, just with `incsubstute=''`? Note sure what to do about `any "live" handling in general` though.

justinmk commented:
> Ok we have some basics: 561c1e3#diff-4dbe255ebfc94f6664c66ffeab750809R83

Confusingly, the `helpers.feed()` method is not equivalent to VimL `feedkeys()`. `helpers.feed()` simulates actual user input. Where I would like to verify that if plugins call `:substitute` or `feedkeys(':s...')`, incsub should not be invoked. Easy way to verify this is to check if the `incsub` buffer exists after either of those scenarios. (With this PR, the `incsub` buffer hangs around as an unlisted buffer.)

justinmk commented:
Woo, ASAN looks good. Going to wrap this up now.

KillTheMule committed:
'inccommand': test: scripts/feedkeys() should not trigger preview

KillTheMule commented:
> highlight bug #5226 (comment)

Note that this is a highlighting bug for single-line substitutions containing escaped chars, but makes the whole thing totally break down with multiline matches.

justinmk commented:
Damn, that's disappointing. I would guess that multiline searches aren't common for many users, but it's still a bad bug.

KillTheMule commented:
It still works as a substitution, just the preview is borked :) But it also means the data structure needs to be reworked, because right now it can only save one line that contains the whole match.

justinmk committed:
'inccommand': preserve b:changedtick

KillTheMule commented on test/functional/ui/incsubstitute_spec.lua:
```diff
@@ -0,0 +1,1286 @@
+      end
+  end)
+
+  it('work then mappings move the cursor', function()
```
Small typo there: then -> when

justinmk committed:
'inccommand': preserve 'modified'

During the live preview, the buffer-local 'modified' flag
should not be changed.

justinmk committed:
'inccommand': disable 'cursorline', 'spell' in preview

justinmk committed:
'inccommand': set buffer name to [Preview]

[inc_sub] is less obvious for users. Also, in the future we may want to
generalize the idea of a "preview buffer", or "incremental commands"
besides :substitute.

justinmk committed:
'inccommand': format line numbers as "|123| "

This matches what Quickfix traditionally does.

justinmk committed:
'inccommand': rename 'incsubstitute'

'inccommand' allows us to expand the feature to other commands, such as:
    :cdo
    :cfdo
    :global

Also rename "IncSubstitute" highlight group to "Substitute".

justinmk commented:
Renamed `'incsubstitute'` to `'inccommand'` so that we can expand it to other commands besides `:substitute` without adding new options (that was already implied in the implementation, re `eap->is_live`).

justinmk committed:
'inccommand': Do not trigger during scripts, feedkeys().

KillTheMule commented on test/functional/ui/incsubstitute_spec.lua:
```diff
@@ -52,6 +52,31 @@ local function common_setup(screen, inccommand, text)
+  it("when invoked by feedkeys() in a script ", function()
+    source(':call feedkeys(":%s/tw/MO/g\\<CR>")')
+    wait()
+    eq(1, eval("bufnr('$')"))
```
Are you sure this checks what we want? By putting `\<CR>` at the end of the source, the command would finish, so the preview window is supposed to be closed at this point anyways. Maybe it's useful to check for it, but you can't be sure it didn't show inbetween,  can you?

justinmk commented:
> Are you sure this checks what we want? By putting `<CR>` at the end of the source, the command would finish, so the preview window is supposed to be closed at this point anyways.

@KillTheMule The compromise made in this PR to avoid the special `buflist_new()` and `do_ecmd()` special-cases, is to keep the buffer around as an unlisted buffer. (That's more or less idiomatic, and less prone to bugs.)

Checking `bufnr('$')` is sufficient because `bufnr('$')` includes unlisted buffers. Note that before https://github.com/neovim/neovim/pull/5561/commits/48a986c030a8cb0578547f312b2b7d56dc73cf15 the test correctly fails.

justinmk commented:
@bfredl @KillTheMule I added a [new test](https://github.com/neovim/neovim/pull/5561/commits/b0b8b277eb92634af6f65f5fd7585541a2e0813a#diff-74648f2c6258b7cd7be5e1d661b3a2ebR141) which compares `undotree()` before and after a canceled `:sub` preview. Here's the failure:

```
:substitute, 'inccommand' preserves undotree()
test/functional/ui/incsubstitute_spec.lua:164: Expected objects to be the same.
Passed in:
(table) {
 *[entries] = {
    [1] = {
      [seq] = 1
      [time] = 1478533735 }
    [2] = {
      [alt] = { ... more }
      [seq] = 3
      [time] = 1478533735 }
    [3] = {
      [seq] = 4
      [time] = 1478533735 }
   *[4] = {
     *[newhead] = 1
      [seq] = 5
      [time] = 1478533735 }
    [5] = {
      [curhead] = 1
      [seq] = 6
      [time] = 1478533735 }
    [6] = {
      [seq] = 7
      [time] = 1478533735 } }
  [save_cur] = 0
  [save_last] = 0
  [seq_cur] = 5
  [seq_last] = 7
  [synced] = 1
  [time_cur] = 1478533735 }
Expected:
(table) {
 *[entries] = {
    [1] = {
      [seq] = 1
      [time] = 1478533735 }
    [2] = {
      [alt] = { ... more }
      [seq] = 3
      [time] = 1478533735 }
    [3] = {
      [seq] = 4
      [time] = 1478533735 }
   *[4] = {
      [seq] = 5
      [time] = 1478533735 }
    [5] = {
      [curhead] = 1
      [seq] = 6
      [time] = 1478533735 }
    [6] = {
      [newhead] = 1
      [seq] = 7
      [time] = 1478533735 } }
  [save_cur] = 0
  [save_last] = 0
  [seq_cur] = 5
  [seq_last] = 7
  [synced] = 1
  [time_cur] = 1478533735 }
```

The difference is that `[newhead] = 1` on entry 6 (expected) vs entry entry 4 (actual). I don't understand `u_undo_and_forget()` at all, so I'm hoping you have some insight. Is this expected?

justinmk committed:
perf: do_sub(): avoid work, avoid screen updates

- Don't fill matched_lines if not showing a preview (!eap->is_live).
- Encapsulate: Move cursor placement logic to show_sub().

justinmk committed:
'inccommand': Preserve curbuf->b_u_newhead.

Add tests for undotree().

Helped-by: Björn Linse <bjorn.linse@gmail.com>

When "curhead" points to a valid head, the value of "newhead" is
meaningless (and really should be set to null). The undo state for
a buffer is _logically_ the enum:

  enum UndoState {
    CurrentHead(head),
    NewHead(head),
    EmptyTree
  }

nvim _represents_ this as: whenever `curbuf->b_u_curhead` is nonnull it
should be used as the current head, and `curbuf->b_u_newhead` is
ignored. If the there is a current head, then this will be redoed on the
next redo, and its parent will be undone on next undo. Only if
`b_u_curhead` is NULL, `b_u_newhead` will be used as the head to undo
(and it is not possible to redo). Also both can be NULL, to indicate an
empty undotree. (To be fair, this only strictly true when calling undo.c
from the outside, in some places _within_ a function in undo.c both
values might be meaningful)

Apparently `undotree()` breaks this non-abstraction, this _cosmetic_
issue can easily be fixed by `ex_substitute` also saving and restoring
`b_u_newhead`, but is doesn't reflect any error really how
`u_undo_and_forget` manipulates the _actual_ state of the undo tree.

KillTheMule commented:
I don't fully understand it either, but I think this is not expected. Could you rerun the test with `feed([[:%s/e<C-\><C-N>]])` (or if this does not fail, add keys til the first failure to produce a minimal example)? I'd have a shot at going through the logic then.

bfredl commented:
Here is the thing. Whenever "curhead" points to a valid head, the value of "newhead" is meaningless (and really should be set to null). The undo state for a buffer is _logically_ the enum:
```
enum UndoState {
    CurrentHead(head),
    NewHead(head),
    EmptyTree
}
```
nvim _represents_ this as, whenever `curbuf->b_u_curhead` is nonnull it should be used as the current head, and `curbuf->b_u_newhead` is ignored.If the there is a current head, then this will be redoed on the next redo, and its parent will be undone on next undo.
Only if `b_u_curhead` is NULL, `b_u_newhead` will be used as the head to undo (and it is not possible to redo). Also both can be NULL, to indicate an empty undotree. (To be fair, this only strictly true when calling undo.c from the outside, in some places _within_ a function in undo.c both values might be meaningful) 

Apparently `undotree()` breaks this non-abstraction, this _cosmetic_ issue can easily be fixed by `ex_substitute` also saving and restoring `b_u_newhead`, but is doesn't reflect any error really how `u_undo_and_forget` manipulates the _actual_ state of the undo tree.

justinmk commented:
Thank you @bfredl , see https://github.com/neovim/neovim/pull/5561/commits/e99890f3d6ea720dc25bab722978ab2f9af50d59

justinmk committed:
Merge #5561 'inccommand'

Initial work by:
  Robin Elrharbi-Fleury (Robinhola)
  Audrey Rayé (Adrey06)
  Philémon Hullot (DesbyP)
  Aymeric Collange (aym7)
  Clément Guyomard (Clement0)

Major revisions by:
  KillTheMule
  Björn Linse <bjorn.linse@gmail.com>
  Justin M. Keyes <justinkz@gmail.com>

justinmk commented:
Merging this, but I think the cursor movement should be revisited. The "jumping around" while building the `:sub` command is not really helpful. Especially with `inccommand=split` where results are shown without moving the cursor at all.

Alok commented:
Could this be used to extend the `incsearch` option to highlight *all* candidate matches rather than just the current one (I think `emacs`does this)?

justinmk commented:
@Alok that's being discussed on vim_dev so we'll wait for it to land there.

gaving commented:
Congrats on the merge of this rockin' feature, this [and the associated PR's] are a great example of OSS done right!

xdxter commented:
Thanks for the work! This feature **rocks**!!

Though, I wonder how hard would it be to make it work with the ```:Subvert``` command from tpope's [abolish](https://github.com/tpope/vim-abolish) plugin, which is immensely helpful in real-life substitution use cases.

dsifford commented:
@justinmk 

> that's being discussed on vim_dev so we'll wait for it to land there.

You wouldn't happen to have a link to that discussion thread and/or an update on the status of that, would you?

Thanks in advance!

justinmk commented:
@dsifford It's already implemented in Vim and Nvim 0.3: https://github.com/neovim/neovim/pull/7463

dsifford commented:
@justinmk Is there a visual example of this somewhere I can take a look at? I'm thinking that maybe what I'm expecting is different than what's already implemented. I've got both `hlsearch` and `incsearch` enabled and it's still not doing what I'd expect.

justinmk commented:
@dsifford Please always mention the exact version you're using. Also for questions about usage/configuration, try https://vi.stackexchange.com and the resources listed at https://neovim.io/community/.

dsifford commented:
@justinmk Thanks for the info.

I think I may have found a bug in 0.2.2. (everything works as expected in vim proper, but not neovim -- even when running with `-u NONE`).

I'm gonna compile nightly here in a few and test and see if it's already been fixed before opening an issue on this.

justinmk commented:
@dsifford I mentioned in https://github.com/neovim/neovim/pull/5561#issuecomment-383732551 the feature requires Nvim 0.3.

dsifford commented:
D'oh! :confounded: 

I apologize for the noise. Thanks so much for your patience and kind feedback.
